Consequencer updates and first CVecorder PR

gamecat69 opened the pull request:
Note that the doc in the firmware/contrib folder is needed as some old website articles link to this old location e.g. the gearnews article on consequencer.

awonak commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+                        self.CVR[b][i] = [x / 100 if x > 0 else 0 for x in self.CVR[b][i]]
+                        i += 1
+
+                    file.close()
```
nit: this is technically unnecessary since you're using the `with` block, which safely handles closing the file. Same for write.

https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files

gamecat69 replied:
Thanks for the tip! I actually put it in their as I was wondering if micropython handled this differently!

gamecat69 replied:
I have removed both instances of file.close()

awonak commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+                        self.CVR[b][i] = [x / 100 if x > 0 else 0 for x in self.CVR[b][i]]
+                        i += 1
+
+                    file.close()
```
nit: this is technically unnecessary since you're using the `with` block, which safely handles closing the file. Same for write.

https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files

gamecat69 replied:
Thanks for the tip! I actually put it in their as I was wondering if micropython handled this differently!

gamecat69 replied:
Removed

awonak reviewed (approved):
Excellent work on a super fun script! I left a few nitpick and optional comments for improvement. Thanks for making another banger script!

mjaskula commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+        self.bankToSave = 0
+        self.debugTest = False
+
+        self.numCVR = 5  # Number of CV recorder channels - zero based
```
I'm being a bit pedantic here, but the number (count) of something can't really be 0 based. I think you mean 'max' or 'max index'.

mjaskula replied:
Nope, it was just something that slowed me down when reading the code, so I pointed it out as a potential improvement.

mjaskula commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+
+        # B2 short press
+        @b2.handler
+        def b2Pressed():
```
I feel like it might simplify the code a bit to put all the code for button 2 in the falling handler. Then you wouldn't have to do the reversal steps.

gamecat69 replied:
It would simplify the code, but I did it this way by design so that the most used function of the button triggers when the button is pressed and the least used functions trigger then the button is released.

mjaskula commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+                self.bankToSave = self.ActiveBank
+                self.saveState()
+
+    def clearCvrs(self, bank):
```
It unnecessarily complicates the code to have the parameter `bank` be either a string ('all') or an int representing the bank number. Consider instead having a parameter `banks` that was a List containing the banks to clear. In most cases it would be a single item, the bank to clear. In the 'all' case, you would generate the list of all banks (using `range`). But in both cases the code would be the same, just iterate through the list and perform the clear. No more special cases!

gamecat69 replied:
I like this idea, let me have a play with this - thanks!

mjaskula commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+
+    def saveState(self):
+        # generate output filename
+        outputFile = f"saved_state_{self.__class__.__qualname__}_{self.bankToSave}.txt"
```
Why is `bank_to_save` a member variable instead of a method parameter?

gamecat69 replied:
Apologies, I don't understand the question. Would you mind elaborating and also explaining the difference, thanks.

mjaskula replied:
Sorry, typo, it's `bankToSave` not `bank_to_save`

`bankToSave` is a member variable of the class `CVecoder`, that is, it is a field on the class. But I don't think it needs to be. Every time you set `self.bankToSave` you immediately call `self.saveState()`, and you never call `saveState()` without setting `bankToSave` first. They are intimately linked. I don't see any reason why `bankToSave` should be a field on the class, instead it should be a parameter on the `saveState()` method. As in:

```Python
def saveState(self, bankToSave):
    # generate output filename
    outputFile = f"saved_state_{self.__class__.__qualname__}_{bankToSave}.txt"
    
    ...
```

And the call site, at line 181, for example:
```Python
# Save the cleared bank to local storage
self.saveState(b)
```

Does that make sense?

mjaskula replied:
oh, wow, got it. Let's add some documentation to describe this. Maybe at the beginning of the block of vars that are initialized for this purpose in the `__init__`?

gamecat69 replied:
Done! thanks for the suggestion.

mjaskula replied:
New doc looks good.

mjaskula commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+                        break
+            except MemoryError as e:
+                if self.debugTest:
+                    print(f'[{attempts}] Error: Memory allocation failed, retrying: {e}')
```
What changes on the subsequent attempts that allows the allocation to succeed?

gamecat69 replied:
Nothing other than a small passage of time that may have allowed micropython to somehow (and rather hopefully!) fix itself. I do need to clean this section up a little as it is probably no longer needed now I have got to the bottom of the heap fragmentation issue. leave this with me!

mjaskula commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+                    # populate CV recording channel with saved json data
+                    self.CVR[b] = json.loads(jsonData)
+
+                    # convert values in the list from int back to float
```
You do this in two different places. Consider extracting a method.

gamecat69 replied:
I am happy with this being duplicated just twice for the time being. If I end up doing it more than twice I will definitely move it to a method/function.

mjaskula commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+                    for n in range (0, self.stepLength):
+                        self.CVR[b][i].append(0)
+
+    def debugDumpCvr(self):
```
appears unused

gamecat69 replied:
It is. However, I would like to keep it for future use for now - I added a comment to this effect in the code.

gamecat69 commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+            self.bankToSave = b
+            self.saveState()
+
+    def initCvrs(self):
```
I would like to keep the initCvrs function in case I use it again in future versions. Please let me know how you would like me to comment this so you can ignore it in your PR review, thanks.

gamecat69 commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+        # Load CV Recordings from a previously stored state on disk or initialize if blank
+        self.loadState()
+
+        # Test routine, pick a random bank n times and save, then load the state
+        if self.debugTest:
+            print(micropython.mem_info("level"))
+            for n in range(3000):
+                # Clear vars
+                #self.CvRecording = []
+                print(f"Running test: {n}")
+                self.ActiveBank = randint(0, self.numCVRBanks)
+                self.ActiveCvr = randint(0, self.numCVR)
+                for i in range(0, self.stepLength-1):
+                    self.CVR[self.ActiveBank][self.ActiveCvr][i] = uniform(0.0, 9.99)
+                    #print(f"[{self.ActiveBank}][{self.ActiveCvr}][{i}] = {self.CVR[self.ActiveBank][self.ActiveCvr][i]}")
+                self.bankToSave = self.ActiveBank
+                self.saveState()
+                self.loadState()
```
I would like to keep this code exactly where it is please, I do not want to spend more time creating a unit test at this point in time. If this will prevent my pull request so be it.

gamecat69 commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+        # B2 Long press
+        @b2.handler_falling
+        def b2PressedLong():
+            # 2000ms press clears all banks
```
I tried a slightly different approach with CVecorder compared to Consequencer as I wanted the most used function to trigger before the button was released. TBH, it may be slightly easier if there was a long_press(num_seconds) function in the firmware. I feel that this function should also make sure it reverts any change that was made as a result of the short press though too.

gamecat69 commented on software/contrib/consequencer.py:
```diff
@@ -123,12 +134,8 @@ def clockTrigger():
-            if self.clock_step < 128:
-                self.clock_step +=1
-            else:
-                self.clock_step = 0
```
I added it to avoid a growing memory leak as Python just keeps adding more and more memory space to integers that keep growing. However, after testing for a long time I don't believe I need to worry about a memory leak.
Additionally, this simple if statement added latency that I want to keep to minimum.

gamecat69 commented on software/contrib/cvecorder.md:
```diff
@@ -0,0 +1,103 @@
+- Each CV channel records a total of 64 CV values (samples).
+- Two CV samples are recorded for each received gate. One sample is captured on the rising edge, another on the falling edge.
+- Each CV channel is looped with every 32nd gate into the Digital input.
+- Each CV sample is captured with a resolution of 0 - 9.99, providing a resolution of 999 potential CV values to record.
```
Good point! I will make the change,

gamecat69 commented on software/contrib/cvecorder.py:
```diff
@@ -0,0 +1,348 @@
+'''
+
+# Needed if using europi_script
+#class CVecorder(EuroPiScript):
```
Thanks for the info, I will correct this.

gamecat69 commented on software/contrib/consequencer.py:
```diff
@@ -55,12 +55,19 @@ def __init__(self):
+
+        # option to always output a clock on output 4
+        # this helps to sync Consequencer with other modules
+        self.output4isClock = False
```
OK got it, thanks.
Let me have a think about this one. If I can add it to the UI somehow,(preferred) then this suggestion/problem goes away. I may add it to the UI by having a reaaallllyyyyy looooooonnng button press (>4 seconds) or something like that.

mjaskula replied:
Not required for merge.

awonak commented:
It looks like the unit tests are now failing with changes to the menu import, PTAL: https://github.com/Allen-Synthesis/EuroPi/runs/6009089985?check_suite_focus=true

gamecat69 commented:
It looks like it's the "import micropython" line that is causing the
failure.
Kind Regards,
Nik


On Wed, Apr 13, 2022 at 4:26 PM Adam Wonak ***@***.***> wrote:

> It looks like the unit tests are now failing with changes to the menu
> import, PTAL:
> https://github.com/Allen-Synthesis/EuroPi/runs/6009089985?check_suite_focus=true
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Allen-Synthesis/EuroPi/pull/121#issuecomment-1098117537>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACKZSUITPH2JJBO2NAEJZATVE3KQBANCNFSM5S5BNBBQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

mjaskula commented on CHANGELOG.md:
```diff
@@ -1,5 +1,17 @@
+- Consequencer: Added explanation of self.output4isClock to doc
+- CVecorder: Added comments to unused functions which are reserved for future use
+- Cvecorder: doc updates
+- Cvecorder: Added menu handling code
```
A few of these entries are probably more detailed than we want here. This doc should be targeted at the user and paint the broad strokes of what they're getting with an upgrade to the new version. (and here 'user' can mean both the user of a specific script or the user of the API)

So I don't think we need any details on code changes, just user visible functionality changes. Also, CVecoder is new in this release, so we just need an entry saying that it was added.

mjaskula commented:
The only remaining change I'd like to see before merge is the removal of `europi-front.png` (which appears to be unused) and `previous-pcb-design-build-instructions.png`. I discussed my reasoning for the latter file in a previous comment:

>Also, regarding docs/previous-pcb-design-build-instructions.png I'm not sure that an image file is the best way to commit build instructions. It will be very difficult for anyone to make changes to it. Also it references previous designs, which can get confusing. I'd like to suggest removing this file in favor of making any appropriate changes to https://github.com/Allen-Synthesis/EuroPi/blob/main/hardware/build_guide.md

gamecat69 commented:
previous-pcb-design-build-instructions.png and europi-front.png docs removed.
Fix for calibrate.py added.
output4isClock now accessible via the UI.
Added function timing code, but commented out - I will use this in the future.
Removed detail from CVecorder now reads "CVecorder: Added"

awonak commented:
Thanks for sharing another amazing script!
